presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java

```diff
@@ -0,0 +1,235 @@
+            this.stats = requireNonNull(stats, "stats is null");
+            this.lookup = requireNonNull(lookup, "lookup is null");
+        }
+
+        @Override
+        protected PlanNodeCostEstimate visitPlan(PlanNode node, Void context)
+        {
+            return UNKNOWN_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitGroupReference(GroupReference node, Void context)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitOutput(OutputNode node, Void context)
+        {
+            return ZERO_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitFilter(FilterNode node, Void context)
+        {
+            return cpuCost(getStats(node.getSource()).getOutputSizeInBytes());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitProject(ProjectNode node, Void context)
+        {
+            return cpuCost(getStats(node).getOutputSizeInBytes());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitAggregation(AggregationNode node, Void context)
+        {
+            PlanNodeStatsEstimate aggregationStats = getStats(node);
+            PlanNodeStatsEstimate sourceStats = getStats(node.getSource());
+            return new PlanNodeCostEstimate(sourceStats.getOutputSizeInBytes(), aggregationStats.getOutputSizeInBytes(), 0);
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitJoin(JoinNode node, Void context)
+        {
+            return calculateJoinCost(
+                    node,
+                    node.getLeft(),
+                    node.getRight(),
+                    node.getDistributionType().orElse(JoinNode.DistributionType.PARTITIONED).equals(JoinNode.DistributionType.REPLICATED));
+        }
+
+        private PlanNodeCostEstimate calculateJoinCost(PlanNode join, PlanNode probe, PlanNode build, boolean replicated)
+        {
+            int numberOfNodesMultiplier = replicated ? numberOfNodes : 1;
+
+            PlanNodeStatsEstimate probeStats = getStats(probe);
+            PlanNodeStatsEstimate buildStats = getStats(build);
+            PlanNodeStatsEstimate outputStats = getStats(join);
+
+            double cpuCost = probeStats.getOutputSizeInBytes() +
+                    buildStats.getOutputSizeInBytes() * numberOfNodesMultiplier +
+                    outputStats.getOutputSizeInBytes();
+
+            if (replicated) {
+                // add the cost of a local repartitioning of build side copies
+                // cost of the repartitioning of a single data copy has been already added in calculateExchangeCost
+                cpuCost += buildStats.getOutputSizeInBytes() * (numberOfNodesMultiplier - 1);
+            }
+
+            double memoryCost = buildStats.getOutputSizeInBytes() * numberOfNodesMultiplier;
+
+            return new PlanNodeCostEstimate(cpuCost, memoryCost, 0);
```

Comment:
`replicated` will have always higher cpu and memory, is that desired? Maybe I misunderstood something. If I am correct it may make CBO code easier ;) Decision which join should be used is based mostly on cost of `ExchangeNode`s with an offset from difference between replicated and repartitioned join cost.

Reply:
I don't understand